12.0 remove view seb

seb-odoo opened the pull request:
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:




--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

JKE-be commented:
louch to be in base

seb-odoo commented:
I tested myself both theme and normal view, including with inherit for force unlink, and also added an actual test case for comparison.

I don't see where else to put this fix than base. Other use cases can be found for this issue outside of website: any copied view in odoo which uses data/fields/... only present on its own module will crash if that module is removed. It just so happens that copying a view is rare in general but is a standard procedure for multiwebsite.

I'm just a bit afraid of potential side effects that I'm unaware of.

Other than this concern for me this can be reviewed.

rdeodoo commented on addons/website_theme_install/models/ir_module_module.py:
```diff
@@ -236,7 +236,10 @@ def _theme_cleanup(self, model_name, website):
-        orphans = model.with_context(active_test=False).search([
+        # use active_test to also unlink archived models, and use _force_unlink
+        # to also unlink inherited models
+        orphans = model.with_context(active_test=False, _force_unlink=True).search([
```
What happened to MODULE_UNINSTALL_FLAG?
Could not use it outside the `base ir_model` file?

seb-odoo replied:
I was hesitant because it wasn't used in other places but I think it's a good idea. So I'll say I was just waiting for you to ask before adding it.

rdeodoo commented on addons/website_sale_comparison/tests/test_website_sale_comparison.py:
```diff
@@ -0,0 +1,30 @@
+        # first test that the comparison button is present initially and
+        # actually create a copy of the view by using the editor to trigger COW
+        # on the product page and all its tree (including the view we want)
+        self.phantom_js("/", "odoo.__DEBUG__.services['web_tour.tour'].run('tour_shop_product_page_comparison')", "odoo.__DEBUG__.services['web_tour.tour'].tours.tour_shop_product_page_comparison.ready", login="admin")
```
Can't all this be done in python?
```python
# Retrieve the generic view
generic_view = View.with_context(website_id=None).get_view_id('website_sale.product')
# Trigger COW
generic_view.with_context(website=1).write({'name': 'Trigger COW'})
```
Then remove the module

And next JS tour can simply be replaced by

```python
# Get specific view DOM
specific_view = View.with_context(website_id=1).get_view_id('website_sale.product')
specific_view_arch = View.browse(specific_view).with_context(load_all_views=True, website_id=1).read_combined(['arch'])['arch']
# Check compare is not in it anymore
self.assertEqual('compare' in specific_view_arch, False, "There should not be compare stuff left as we uninstalled the module")
```

seb-odoo replied:
Well I think the question here is do we want to test an actual flow, or do we want to unit test the method?

The test I wrote is almost exactly matching what I did functionally to get the issue, which I think it is nice to cover.

But I would also be in favor of unit testing the method by adding a test more or less like you suggest in base.

rdeodoo replied:
Honestly I feel like it would be better to test the uninstall module (triggering cow & checking after uninstall if compare is not there anymore) in python like I suggested.

And add a test to test comparison if needed but keep both tests separated.

Would be clearer that way, each test testing its own stuff. And in the future if we need to change the comparison tour (make it bigger, test new stuff etc) we wont have the view uninstall during removal at the same time

seb-odoo replied:
Ok as we saw together, we adapted those snippets and we'll keep them instead.

seb-odoo commented:
I  updated the PR with a more ideal test, including inherit, and also added few more small commits.

This can be review again.

JKE-be commented on addons/website/models/website.py:
```diff
@@ -527,9 +527,9 @@ def viewref(self, view_id, raise_if_not_found=True):
+                view = self.env.ref(view_id, raise_if_not_found=raise_if_not_found)
                 # self.env.ref might return something else than an ir.ui.view (eg: a theme.ir.ui.view)
-                if view._name != 'ir.ui.view':
+                if not view or view._name != 'ir.ui.view':
```
if view and view._name...

seb-odoo replied:
`not view or` ensure `view=None` when it's falsy
`view and` the view will keep any falsy value it had

It doesn't actually matter in this case since we only use that value in a condition after.

JKE-be replied:
Si tu n'as pas a manger devant toi ou que ce sont des légumes, alors vide ton assiette (ne fait rien)...
(qui est deja Falsy et retournait donc None)

Ca a quand meme plus de sens de dire 
"Si tu as a manger, et que ce que tu manges n'est pas des légumes, vide l'assiette "

mais mtnt que on ensure que si pas de vue on retourne un recordset vide, c ok comme ca ;)

seb-odoo commented on addons/website/models/website.py:
```diff
@@ -527,9 +527,9 @@ def viewref(self, view_id, raise_if_not_found=True):
                 view = views.filter_duplicate()
             else:
-                view = self.env.ref(view_id)
+                view = self.env.ref(view_id, raise_if_not_found=raise_if_not_found)
```
It doesn't actually matter if it's `raise_if_not_found=False` or `raise_if_not_found=raise_if_not_found` because if the view is empty, there is another `if raise_if_not_found` later on the method.

We just need to decide if we want in that case to raise the original exception from ref or the new exception from viewref.

What is definitely not correct is to pass True (default value) like before.

seb-odoo commented:
Updated the `viewref` fix commit to always set `raise_if_not_found=False` and also fixed another issue where it would return None instead of a recordset.

Can be reviewed again

JKE-be commented on addons/website/models/website.py:
```diff
@@ -521,20 +527,20 @@ def viewref(self, view_id, raise_if_not_found=True):
-            return view
-        if raise_if_not_found:
+        if not view and raise_if_not_found:
             raise ValueError('No record found for unique ID %s. It may have been deleted.' % (view_id))
```
It will be raised here !

xmo-odoo commented:
ping @Elkasitu that's uninstall stuff, you might have better insight as to what might break or be missing

JKE-be commented:
@Elkasitu 

If nothing horrible here for you, can you ask to robodoo to merge please.
Thank you

@robodoo delegate=Elkasitu

seb-odoo commented:
Thank you for checking it up.

All of those commits are independent (except maybe the last 2), I just put them all here to not bother with 5 different PRs, so I prefer fast forward.

Elkasitu commented:
@robodoo r+ rebase-ff please and thank you

robodoo commented:
I'm sorry, @Elkasitu. You can't method=rebase-ff.

Elkasitu commented:
@robodoo r-

Elkasitu commented:
@JKE-be looks like delegate works for r+ but not for setting the method :/

seb-odoo commented:
JKE won't be here today, maybe we can ask help from @xmo-odoo ?

xmo-odoo commented:
robodoo rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward

Elkasitu commented:
robodoo r+

Elkasitu commented:
robodoo retry

robodoo commented:
Merged, thanks!
